libraries/abstractions/ble_hal/test/src/iot_test_ble_hal_integration.c

```diff
@@ -50,4 +77,245 @@ TEST_TEAR_DOWN( Full_BLE_Integration_Test )
+ * Make sure stack state is enable (callback received) no later than 2.5 seconds after pxEnable returns*/
+TEST( Full_BLE_Integration_Test, BLE_Enable_Disable_BT_Module )
+{
+    BTStatus_t xStatus = eBTStatusSuccess;
+    BLETESTInitDeinitCallback_t xInitDeinitCb;
+    clock_t returnTime, cbRecvTime;
+
+    /* disable */
+    prvBLEEnable( false );
+
+    /* enable */
+    xStatus = g_pxBTInterface->pxEnable( 0 );
+    returnTime = clock();
+    TEST_ASSERT_EQUAL( eBTStatusSuccess, xStatus );
+
+    xStatus = prvWaitEventFromQueue( eBLEHALEventEnableDisableCb, NO_HANDLE, ( void * ) &xInitDeinitCb, sizeof( BLETESTInitDeinitCallback_t ), BLE_TESTS_WAIT );
+    cbRecvTime = clock();
+    TEST_ASSERT_EQUAL( eBTStatusSuccess, xStatus );
+    TEST_ASSERT_EQUAL( eBTstateOn, xInitDeinitCb.xBLEState );
+    TEST_ASSERT_LESS_THAN( CLOCKS_PER_SEC * 5, ( cbRecvTime - returnTime ) * 2 );
+}
+
+
+/* Crash if calling pxEnable twice (MTK)
+ * (1)init -> (2)enable -> (3)deinit -> (4)init -> (5)enable
+ * There are 2 issues with this sequence:
+ * (4)init reset stack state to disabled even though it's still enabled
+ * (5)enable trigger pxEnable again while MTK stack is enabled -> mtk crashed   */
+TEST( Full_BLE_Integration_Test, BLE_Init_Enable_Twice )
+{
+    prvGAPInitEnableTwice();
+}
+
+/*Advertisement interval measured OTA can be out the range set by app, after reset BT stack, adv interval can change to 1.28s.
+ * Make sure KPI is consistent after reset BT.*/
+TEST( Full_BLE_Integration_Test, BLE_Advertise_Interval_Consistent_After_BT_Reset )
+{
+    /* First time connection disconnects. Got First KPI. */
+    prvWaitConnection( false );
+
+    /* BT reset. */
+    BTStatus_t xStatus = eBTStatusSuccess;
+    BLETESTInitDeinitCallback_t xInitDeinitCb;
+
+    prvBTUnregister();
+    prvBLEEnable( false );
+
+    xStatus = g_pxBTInterface->pxBtManagerCleanup();
+    TEST_ASSERT_EQUAL( eBTStatusSuccess, xStatus );
+
+    xStatus = g_pxBTInterface->pxBtManagerInit( &g_xBTManagerCb );
+    TEST_ASSERT_EQUAL( eBTStatusSuccess, xStatus );
+
+    prvBLEEnable( true );
+    prvBLEGAPInit();
+    prvBLEGATTInit();
+    prvCreateAndStartServiceB();
+    prvSetAdvProperty();
+    prvSetAdvData();
+
+    /* Second time connection begins. Got second KPI. */
+    prvStartAdvertisement();
+
+    prvWaitConnection( true );
+
+    /* Result is on RPI. Write it back to device. */
+    prvGetResult( bletestATTR_SRVCB_CHAR_D,
+                  false,
+                  0 );
+}
+
+/* If data size is > MTU - 3 then BT stack can truncate it to MTU - 3 and keep trying to send it over to other peer.
+ * Make sure calling pxSendIndication() with xLen > MTU - 3 and stack returns failure.*/
+TEST( Full_BLE_Integration_Test, BLE_Write_Notification_Size_Greater_Than_MTU_3 )
+{
+    BTStatus_t xStatus, xfStatus;
+    BLETESTindicateCallback_t xIndicateEvent;
+    uint8_t ucLargeBuffer[ bletestsMTU_SIZE1 + 2 ];
```

Comment:
Be careful allocating an array like this on the stack. It should be ok but it is risky. Make it static.